### PR TITLE
move extension to tools and merge build and publish

### PR DESCRIPTION
**Pull request by 928PJY**
move extension to tools and merge build with publish

### CODE REVIEW

File: tools/buildDocfxPreviewExtension.ps1 [hunk 1]
```diff
@@ -0,0 +1,44 @@
+param(
+    [string] $command = "package",
+    [string] $version
+)
+
+################################################################################################
+# Usage:
+# Run buildDocfxPreviewExtension.ps1
+#   [-command Command]: package or publish, default to be package
+#   [-version Version]: Publish version whichc should newer than published version
```
Comment (superyyrrzz) [inline tools/buildDocfxPreviewExtension.ps1]: >c [](start = 45, length = 1)

remove

Comment (superyyrrzz) [inline tools/buildDocfxPreviewExtension.ps1]: be newer than

**Comment by superyyrrzz** [comment]
:shipit:

File: tools/VscPreviewExtension/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+## Known Issues
+* Preview match: File inclusion
+* Match when multiple open editor
```
Comment (hellosnow) [inline tools/VscPreviewExtension/CHANGELOG.md]: >Match when multiple open editor [](start = 2, length = 31)

more detailed information to describe the know issue

File: tools/buildDocfxPreviewExtension.ps1 [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    [string] $command = "package",
+    [string] $version
+)
+
+################################################################################################
+# Usage:
+# Run buildDocfxPreviewExtension.ps1
+#   [-command Command]: package or publish, default to be package
+#   [-version Version]: Publish version which should be newer than published version
+################################################################################################
```
Comment (vicancy) [inline tools/buildDocfxPreviewExtension.ps1]: move into VscPreviewExtension and rename to build.ps1

File: tools/VscPreviewExtension/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+## Current Version **0.0.21**
```
Comment (vicancy) [inline tools/VscPreviewExtension/CHANGELOG.md]: Move to top? How about:

# Current Version...
## New Features
...
## Known Issues
...

## Version History
...

File: tools/buildDocfxPreviewExtension.ps1 [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+Write-Host "Build $HttpServiceFolder to the target folder"
+$outputFolder = Join-Path $scriptHome "$extensionFolder\$HttpServiceFolder"
+& dotnet publish ".\$HttpServiceFolder" -c $configuration -f $framework -o $outputFolder
+ProcessLastExitCode $lastexitcode "Error occurs when building $HttpServiceFolder"
+Pop-Location
+
+$extensionFolderPath = Join-Path $scriptHome $extensionFolder
+Push-Location $extensionFolderPath
+
+Write-Host "`n$command extension"
```
Comment (vicancy) [inline tools/buildDocfxPreviewExtension.ps1]: >`n [](start = 12, length = 2)

what does `` `n `` mean?

Comment (928PJY) [inline tools/buildDocfxPreviewExtension.ps1]: a new line

---
In reply to: [109083573](https://github.com/dotnet/docfx/pull/1511#discussion_r109083573) [](ancestors = 109083573)

File: tools/buildDocfxPreviewExtension.ps1 [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+$configuration = "Release"
+$framework = "net46"
+$HttpServiceFolder = "DfmHttpService"
+$extensionFolder = "VscPreviewExtension"
+
+function ProcessLastExitCode {
+    param($exitCode, $msg)
+    if ($exitCode -ne 0) {
+        Write-Error "$msg, exit code: $exitCode"
+        Pop-Location
```
Comment (hellosnow) [inline tools/buildDocfxPreviewExtension.ps1]: >Pop-Location [](start = 8, length = 12)

did you push-location before pop?

Comment (928PJY) [inline tools/buildDocfxPreviewExtension.ps1]: Line 29 and Line 38

---
In reply to: [109084783](https://github.com/dotnet/docfx/pull/1511#discussion_r109084783) [](ancestors = 109084783)

Comment (hellosnow) [inline tools/buildDocfxPreviewExtension.ps1]: ok. I reviewed the function first... missing the later push

---
In reply to: [109084783](https://github.com/dotnet/docfx/pull/1511#discussion_r109084783) [](ancestors = 109084783)

File: tools/VscPreviewExtension/build.ps1 [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#   [-command Command]: package or publish, default to be package
+#   [-version Version]: Publish version which should be newer than published version
+################################################################################################
+$ErrorActionPreference = 'Stop'
+$scriptPath = $MyInvocation.MyCommand.Path
+$scriptHome = Split-Path $scriptPath
+$configuration = "Release"
+$framework = "net46"
+$HttpServiceFolder = "DfmHttpService"
+$extensionFolder = "VscPreviewExtension"
```
Comment (vicancy) [inline tools/VscPreviewExtension/build.ps1]: >$extensionFolder [](start = 0, length = 16)

not used

File: tools/VscPreviewExtension/build.ps1 [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        Pop-Location
+        Exit 1
+    }
+}
+
+Push-Location $scriptHome
+
+Write-Host "Build $HttpServiceFolder to the target folder"
+$outputFolder = Join-Path $scriptHome $HttpServiceFolder
+& dotnet publish "..\$HttpServiceFolder" -c $configuration -f $framework -o $outputFolder
```
Comment (vicancy) [inline tools/VscPreviewExtension/build.ps1]: >"..\$HttpServiceFolder" [](start = 17, length = 23)

use a var

**Comment by vicancy** [comment]
:shipit:

File: tools/VscPreviewExtension/build.ps1 (hunk 7 above)
Comment (hellosnow) [inline tools/VscPreviewExtension/build.ps1]: >$extensionFolder [](start = 0, length = 16)

not used?

**Comment by hellosnow** [comment]
:shipit:

File: tools/VscPreviewExtension/build.ps1 [hunk 9]
```diff
@@ -15,7 +15,7 @@ $scriptHome = Split-Path $scriptPath
 $configuration = "Release"
 $framework = "net46"
 $HttpServiceFolder = "DfmHttpService"
-$extensionFolder = "VscPreviewExtension"
+$HttpServiceFolderPath = "..\$HttpServiceFolder"
```
Comment (vicancy) [inline tools/VscPreviewExtension/build.ps1]: naming convention: lowercase first letter

File: tools/VscPreviewExtension/.vscodeignore [hunk 10]
```diff
@@ -6,4 +6,6 @@ src/**
 **/*.map
 .gitignore
 tsconfig.json
-img/**
\ No newline at end of file
+img/**
+*.cmd
```
Comment (vicancy) [inline tools/VscPreviewExtension/.vscodeignore]: restrict to build.cmd?

File: tools/VscPreviewExtension/build.ps1 (hunk 7 above)
Comment (928PJY) [inline tools/VscPreviewExtension/build.ps1]: known issue, resolved.

---
In reply to: [109098566](https://github.com/dotnet/docfx/pull/1511#discussion_r109098566) [](ancestors = 109098566)

**Pull request merged**